src/Node.hs

```diff
@@ -69,17 +89,28 @@ data ConversationActions m = ConversationActions {
        close :: m ()  -- TODO: needed? if so only for early close. Should by automatic.
      }
 
+type ListenerIndex m = Map MessageName (ListenerAction m)
+
+makeListenerIndex :: [Listener m] -> (ListenerIndex m, [MessageName])
+makeListenerIndex = foldr combine (M.empty, [])
+    where
+    combine (Listener name action) (map, existing) =
+        let (replaced, map') = M.insertLookupWithKey (\_ _ _ -> action) name action map
+            overlapping = maybe [] (const [name]) replaced
+        in  (map', overlapping ++ existing)
+
 startNode
     :: forall m g .
        ( Mockable Fork m, Mockable RunInUnboundThread m, Mockable Throw m
-       , Mockable Channel m, Mockable SharedAtomic m, RandomGen g )
+       , Mockable Channel m, Mockable SharedAtomic m, RandomGen g
+       , MonadFix m )
     => NT.Transport m
     -> g
     -> [Worker m]
     -> [Listener m]
     -> m (Node m)
-startNode transport prng workers listeners = do
-    node <- LL.startNode transport prng handlerIn handlerInOut
+startNode transport prng workers listeners = mdo
+    node <- LL.startNode transport prng (handlerIn node sendAction) (handlerInOut node)
```

Comment:
Why do we need to use `RecursiveDo` here and pass `node` in `hadlerInOut` if this `node` isn't used in that handler now? Or we just probably may want to use it there?